File: client/lib/services/household_service.dart
import 'package:sambo/models/auth_user.dart';
import 'package:sambo/models/chore.dart';
import 'package:sambo/models/household.dart';
import 'package:sambo/models/household_membership.dart';
import 'package:sambo/services/api_client.dart';
import 'package:sambo/services/auth_service.dart';

class HouseholdService {
  HouseholdService._();
  static final HouseholdService instance = HouseholdService._();

  Future<Household> get() async {
    final json = await ApiClient.instance.getJson('/api/household');
    return Household.fromJson(json);
  }

  Future<Household> rename(String name) async {
    await ApiClient.instance.putJson('/api/household', body: {'name': name});
    return get();
  }

  Future<List<UserSummary>> members() async {
    final res = await ApiClient.instance.getJsonList('/api/household/members');
    return res
        .map((e) => UserSummary.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<List<HouseholdMembership>> memberships() async {
    final res =
        await ApiClient.instance.getJsonList('/api/household/memberships');
    return res
        .map((e) => HouseholdMembership.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  /// Switch the active household. Backend mints a fresh JWT; we swap it in so
  /// every subsequent request scopes to the new tenant.
  Future<void> switchActive(String householdId) async {
    final json = await ApiClient.instance
        .postJson('/api/household/switch', body: {'householdId': householdId});
    await _applySession(json);
  }

  /// Leave a household. If it was the active one and the user has another
  /// household, the response carries a fresh JWT for the fallback. If they
  /// just left their last household, the server returns no token — we sign
  /// out so the user lands back on the login screen.
  Future<void> leave(String householdId) async {
    final json = await ApiClient.instance
        .postJson('/api/household/leave', body: {'householdId': householdId});
    final token = json['accessToken'] as String?;
    if (token == null) {
      await AuthService.instance.signOut();
      return;
    }
    await _applySession(json);
  }

  /// Create a new household with the caller as ADMIN, automatically becoming
  /// the new active household.
  Future<void> create(String name) async {
    final json =
        await ApiClient.instance.postJson('/api/household', body: {'name': name});
    await _applySession(json);
  }

  Future<void> _applySession(Map<String, dynamic> json) async {
    final token = json['accessToken'] as String;
    final user = AuthUser.fromJson(json['user'] as Map<String, dynamic>);
    await AuthService.instance.setSession(token, user);
  }
}

File: server/src/main/java/com/sambo/household/HouseholdService.java
package com.sambo.household;

import com.sambo.auth.dto.AuthUserDto;
import com.sambo.auth.jwt.JwtService;
import com.sambo.auth.jwt.SamboPrincipal;
import com.sambo.household.dto.HouseholdMembershipDto;
import com.sambo.household.dto.HouseholdSessionResponse;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Comparator;
import java.util.List;
import java.util.UUID;

/**
 * Membership-level operations on households. Tenant scoping is enforced by
 * always reading the authenticated principal — callers cannot act on a
 * household they aren't a member of.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class HouseholdService {

    private final AppUserRepository userRepo;
    private final HouseholdRepository householdRepo;
    private final HouseholdMembershipRepository membershipRepo;
    private final JwtService jwtService;

    /** Every household the user belongs to, with role + active flag. */
    @Transactional(readOnly = true)
    public List<HouseholdMembershipDto> listMemberships(SamboPrincipal principal) {
        UUID activeId = principal.householdId();
        // JOIN FETCH avoids N+1 on the household side.
        return membershipRepo.findByUserIdFetchingHousehold(principal.userId()).stream()
            .map(m -> HouseholdMembershipDto.from(m,
                activeId != null && activeId.equals(m.getHousehold().getId())))
            .sorted(Comparator.comparing(HouseholdMembershipDto::joinedAt))
            .toList();
    }

    /**
     * Switch the user's active household. The target must already be one of
     * their memberships — there's no implicit join here. Returns a fresh JWT
     * scoped to the new active household + role.
     */
    @Transactional
    public HouseholdSessionResponse switchActive(UUID targetHouseholdId, SamboPrincipal principal) {
        AppUser user = userRepo.findById(principal.userId())
            .orElseThrow(() -> new EntityNotFoundException("User not found"));

        HouseholdMembership membership = membershipRepo
            .findByUserIdAndHouseholdId(user.getId(), targetHouseholdId)
            .orElseThrow(() -> new AccessDeniedException(
                "Du är inte medlem i det hushållet"));

        user.setActiveHousehold(membership.getHousehold());
        userRepo.save(user);

        String jwt = jwtService.issue(user, membership);
        log.info("User {} switched active household to {}", user.getEmail(), targetHouseholdId);
        return new HouseholdSessionResponse(jwt, AuthUserDto.from(user, membership));
    }

    /**
     * Leave a household. If it was the active one, fall back to another
     * membership (oldest first) — or null if none remain. The household is
     * deleted (with everything that hangs off it) when the last member leaves.
     */
    @Transactional
    public HouseholdSessionResponse leave(UUID targetHouseholdId, SamboPrincipal principal) {
        AppUser user = userRepo.findById(principal.userId())
            .orElseThrow(() -> new EntityNotFoundException("User not found"));

        HouseholdMembership membership = membershipRepo
            .findByUserIdAndHouseholdId(user.getId(), targetHouseholdId)
            .orElseThrow(() -> new AccessDeniedException(
                "Du är inte medlem i det hushållet"));

        membershipRepo.delete(membership);

        boolean wasActive = user.getActiveHousehold() != null
            && user.getActiveHousehold().getId().equals(targetHouseholdId);

        // If no members remain in the household, drop it (cascade purges chores,
        // budget rows, calendar events, etc).
        if (membershipRepo.countByHouseholdId(targetHouseholdId) == 0) {
            householdRepo.deleteById(targetHouseholdId);
        }

        HouseholdMembership fallback = null;
        if (wasActive) {
            fallback = membershipRepo.findByUserId(user.getId()).stream()
                .min(Comparator.comparing(HouseholdMembership::getJoinedAt))
                .orElse(null);
            user.setActiveHousehold(fallback == null ? null : fallback.getHousehold());
            userRepo.save(user);
        }

        log.info("User {} left household {}", user.getEmail(), targetHouseholdId);

        if (fallback == null && wasActive) {
            // No JWT to mint — caller has no active household. Front-end is
            // expected to redirect to a "create or accept invite" screen.
            return new HouseholdSessionResponse(null, AuthUserDto.from(user, null));
        }

        // Either active didn't change (we left a non-active household) — re-mint
        // a JWT for the still-current active membership; or active changed and
        // fallback holds the new one.
        HouseholdMembership active = wasActive
            ? fallback
            : membershipRepo
                .findByUserIdAndHouseholdId(user.getId(), user.getActiveHousehold().getId())
                .orElseThrow();
        String jwt = jwtService.issue(user, active);
        return new HouseholdSessionResponse(jwt, AuthUserDto.from(user, active));
    }

    /**
     * Create a new household with the caller as ADMIN, switch to it, mint a
     * fresh JWT. Subject to the per-user household cap.
     */
    @Transactional
    public HouseholdSessionResponse create(String rawName, SamboPrincipal principal) {
        AppUser user = userRepo.findById(principal.userId())
            .orElseThrow(() -> new EntityNotFoundException("User not found"));

        if (membershipRepo.countByUserId(user.getId()) >= InviteService.MAX_HOUSEHOLDS_PER_USER) {
            throw new TooManyHouseholdsException(
                "Du kan max vara med i " + InviteService.MAX_HOUSEHOLDS_PER_USER + " hushåll. "
                    + "Lämna ett hushåll innan du skapar ett nytt.");
        }

        Household household = householdRepo.save(Household.builder()
            .name(rawName.trim())
            .build());

        HouseholdMembership membership = membershipRepo.save(HouseholdMembership.builder()
            .user(user)
            .household(household)
            .role(Role.ADMIN)
            .build());

        user.setActiveHousehold(household);
        userRepo.save(user);

        String jwt = jwtService.issue(user, membership);
        log.info("User {} created household {} ({})",
            user.getEmail(), household.getId(), household.getName());
        return new HouseholdSessionResponse(jwt, AuthUserDto.from(user, membership));
    }
}
